service/dap/server.go

```diff
@@ -1511,10 +1559,105 @@ func (s *Server) onReverseContinueRequest(request *dap.ReverseContinueRequest) {
 	s.sendNotYetImplementedErrorResponse(request.Request)
 }
 
-// onSetVariableRequest sends a not-yet-implemented error response.
-// Capability 'supportsSetVariable' is not set 'initialize' response.
-func (s *Server) onSetVariableRequest(request *dap.SetVariableRequest) { // TODO V0
-	s.sendNotYetImplementedErrorResponse(request.Request)
+func variableByName(v *proc.Variable, name string) (*proc.Variable, bool) {
+	if v == nil {
+		return nil, false
+	}
+	if v.Name == name {
+		return v, true
+	}
+
+	for _, child := range v.Children {
+		if child.Name == name {
+			return &child, true
+		}
+		// No need to look into the child's children - DAP clients are expected
+		// to provide the variable reference number, so the name is the name of
+		// the variable's direct child.
+	}
+	return nil, false
+}
+
+// nameToEvaluateName finds the variable of the specified name
+// from the given variable and its children, and returns its
+// evaluateName.
+func (s *Server) nameToEvaluateName(v *fullyQualifiedVariable, name string) (string, error) {
+	if v.Name == name && !v.isScope { // will this ever happen?
+		return v.fullyQualifiedNameOrExpr, nil
+	}
+	children, err := s.childrenOfVariables(v)
+	if err != nil {
+		return "", err
+	}
+	evaluateName := name
+	for _, c := range children {
+		if c.Name == name {
+			if c.EvaluateName != "" {
+				evaluateName = c.EvaluateName
+			}
+			break
+		}
+	}
+	return evaluateName, nil
+}
+
+// onSetVariableRequest handles 'setVariable' requests.
+func (s *Server) onSetVariableRequest(request *dap.SetVariableRequest) {
+	arg := request.Arguments
+
+	v, ok := s.variableHandles.get(arg.VariablesReference)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to lookup variable", fmt.Sprintf("unknown reference %d", request.Arguments.VariablesReference))
+		return
+	}
+	// We may need to translate the name to the evaluateName if the name
+	// refers to a field or element of a variable.
+	// https://github.com/microsoft/vscode/issues/120774
+	// Compute the evaluateName by running the same code the onVariablesRequest
+	// handler uses.
+	evaluateName, err := s.nameToEvaluateName(v, arg.Name)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+		return
+	}
+	named, ok := variableByName(v.Variable, arg.Name)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to lookup variable", fmt.Sprintf("unknown reference %d", request.Arguments.VariablesReference))
+		return
+	}
+	// Assume the top-most frame & the current goroutine.
+	// It is not clear if DAP allows SetVariable requests
+	// for variables or memory location not visible from the scode.
```

Comment:
what's "scode"?

I am not sure what DAP allows (since the protocol is not very detailed), but vscode would totally let you set a variable from whatever scope you are viewing in the UI, topmost or not. If for now we are assuming this relative to topmost scope only, we should debug log that we are making this assumption. And lets add a TODO to keep track of the frame for the variable, so we can send more explicit error messages for unsupported frames (like with call) or even allow alternative frames (since I think delve backed allows it).

Reply:
done